hvac/api/auth_methods/cert.py

```diff
@@ -0,0 +1,254 @@
+            'disable_binding': disable_binding,
+        }
+        api_path = '/v1/auth/{mount_point}/config'.format(mount_point=mount_point)
+        return self._adapter.post(
+            url=api_path,
+            json=params,
+        )
+
+    def login(self, name="", cacert=False, cert_pem="", key_pem="", mount_point='cert', use_token=True):
+        """
+        Log in and fetch a token. If there is a valid chain to a CA configured in the method and all role constraints
+            are matched, a token will be issued. If the certificate has DNS SANs in it, each of those will be verified.
+            If Common Name is required to be verified, then it should be a fully qualified DNS domain name and must be
+            duplicated as a DNS SAN
+
+        Supported methods:
+            POST: /auth/<mount point>/login Produces: 200 application/json
+
+        :param name: Authenticate against only the named certificate role, returning its policy list if successful. If
+            not set, defaults to trying all certificate roles and returning any one that matches.
+        :type name: str | unicode
+        :param cacert: The value used here is for the Vault TLS Listener CA certificate, not the CA that issued the
+            client authentication certificate. This can be omitted if the CA used to issue the Vault server certificate
+            is trusted by the local system executing this command.
+        :type cacert: str | bool
+        :param cert_pem: Location of the cert.pem used to authenticate the host.
+        :tupe cert_pem: str | unicode
+        :param key_pem: Location of the public key.pem used to authenticate the host.
+        :param key_pem: str | unicode
+        :param mount_point:
+        :type mount_point:
+        :param use_token: If the returned token is stored in the client
+        :param use_token: bool
+        :return: The response of the login request.
+        :rtype: requests.Response
+        """
+        params = {'use_token': use_token}
+        if name != "":
+            params['name'] = name
+        api_path = '/v1/auth/{mount_point}/login'.format(mount_point=mount_point)
+
+        # Must have cert checking or a CA cert. This is caught lower down but harder to grok
+        if not cacert:
+            # If a cacert is not provided try to drop down to the adapter and get the cert there.
+            # If the cacert is not in the adapter already login will also.
+            if not self._adapter._kwargs.get('verify'):
+                raise self.CertificateAuthError("cacert must be True, a file_path, or valid CA Certificate.")
+            else:
+                cacert = self._adapter._kwargs.get('verify')
+        else:
+            validate_pem_format(cacert, "verify")
+        # if cert_pem is a string its ready to be used and either has the key with it or the key is provided as an arg
+        try:
+            if validate_pem_format(cert_pem, "cert_pem"):
+                tls_update = True
```

Comment:
Erm, this code is unreachable. `validate_pem_format()` is *documented* as returning `bool`, but it actually always returns `None`.

Reply:
Note that the requests library does not accept a PEM string directly s argument -- it must always be a path. So this logic could never have worked anyway.

Reply:
The signature of `validate_pem_format` is `(param_name, param_argument)` but here it's being called like `(param_argument, param_name)` This validation will never pass :(

Reply:
Yes, this was from a miss reading of the requests verify params

:param verify: Either a boolean, in which case it controls whether we verify
            the server's TLS certificate, or a string, in which case it must be a path
            to a CA bundle to use

was read as 

:param verify: Either a boolean, in which case it controls whether we verify,
            the server's TLS certificate, or a string, in which case it must be a path
            to a CA bundle to use
            
I could never figure out why it thought that you could supply the tls cert as an argument but this logic was left in as it wasn't breaking. It wasn't working either.

as @xupefei pointed out I did also transpose the two arguments so this always failed.